MFiles.VAF.Extensions/ExtensionMethods/MFSearchBuilderExtensionMethods/SegmentedSearch.cs

```diff
@@ -0,0 +1,253 @@
+
+			return ForEachSegment(builder,
+				// Note: this method is required because func needs to return the count to be summed by ForEachSegment
+				(vault, conditions) => vault.ObjectSearchOperations.GetObjectCountInSearch(conditions, MFSearchFlags.MFSearchFlagDisableRelevancyRanking),
+				startSegment,
+				segmentLimit,
+				segmentSize);
+		}
+
+		/// <summary>
+		/// Runs a method for each segment in the vault.
+		/// </summary>
+		/// <param name="builder">Search Builder to use for queries.</param>
+		/// <param name="func">Method to be executed for each segment, takes a vault and search conditions and returns count.</param>
+		/// <param name="startSegment">The (zero-based) index of the segment to start at.</param>
+		/// <param name="segmentLimit">The number of total segments to process. See <see cref="DefaultMaximumSegmentIndex"/>.</param>
+		/// <param name="segmentSize">The number of items to include in each segment. See <see cref="DefaultNumberOfItemsInSegment"/>.</param>
+		/// <returns>Total count of objects across vault.</returns>
+		internal static long ForEachSegment(
+			this MFSearchBuilder builder,
+			Func<Vault, SearchConditions, int> func,
+			int startSegment = 0,
+			int segmentLimit = MFSearchBuilderExtensionMethods.DefaultMaximumSegmentIndex,
+			int segmentSize = MFSearchBuilderExtensionMethods.DefaultNumberOfItemsInSegment)
+		{
+			// Sanity.
+			if (null == func)
+				throw new ArgumentNullException(nameof(func));
+			if (startSegment < 0)
+				throw new ArgumentOutOfRangeException(nameof(startSegment), "The start segment must be greater than or equal to zero.");
+			if (segmentSize <= 0)
+				throw new ArgumentOutOfRangeException(nameof(segmentSize), "The segment size must be greater than zero.");
+
+			// Set our start values.
+			var segment = startSegment;
+			long resultCount = 0; // The total number of matched items in all segments.
+
+			// Iterate over segments until we hit the sanity limit,
+			// or until there are no items left to find.
+			while (segment < segmentLimit)
+			{
+				// Clone the search conditions from the supplied builder.
+				var searchConditions = builder.Conditions.Clone();
+
+				// Add a condition for the current segment that we want.
+				searchConditions.Add(-1, SearchConditionSegment(segment, segmentSize));
+
+				// Execute the provided function.
+				// This must return a count of items that were in the current segment,
+				// but it may also execute other code against the items, depending on
+				// what the calling function needs.
+				var searchResultsCount = func(builder.Vault, searchConditions);
+
+				// Remove the condition for the segment.
+				searchConditions.Remove(searchConditions.Count);
+
+				// If we got no items back then we need to check whether a higher segment has items.
+				if (searchResultsCount == 0)
+				{
+					// Clone the original builder conditions.
+					var searchConditionsTopId = builder.Conditions.Clone();

```

Comment:
Could we re-use the existing `searchConditions` variable here instead of creating a new one?

Reply:
Yeah I like this suggestion, see latest commit. Do you know if it is better to clone once outside the loop and keep adding/removing conditions? In most cases I would guess that the clone would be less performant than add and remove.

Reply:
I don't have metrics on performance, but I would tend to agree with you.  Let's go with this for now.